### PR TITLE
Add OnRejected to Request Throttling middleware

### PR DESCRIPTION
**Pull request by Kahbazi**
#10707

### CODE REVIEW

**Comment by anurse** [comment]
Thanks for the submission @Kahbazi ! One of the parts of this middleware I didn't mention clearly is that it is a project our summer intern is working on. However, I think this would be a great opportunity to introduce him to working with external contributors ;).

@DylanDmitri @halter73 Can you both work to see if we can get @Kahbazi 's change iterated on and merged into the repo?

@Kahbazi The first major bit of feedback I can see already is that we're going to need unit tests for this functionality before we could merge it in?

**Comment by Kahbazi** [comment]
@anurse  I'm often looking for bits that I can contribute, and this middleware seems a good one to me. 

> it is a project our summer intern is working on

Does this mean that I shouldn't contribute for this middleware?

**Comment by Tratcher** [comment]
> Does this mean that I shouldn't contribute for this middleware?

You're welcome to contribute but keep two things in mind:
- This component is under active development so be careful not to duplicate work (already assigned issues).
- We'll have our intern @DylanDmitri review and accept the contributions (with guidance from the rest of us).

**Comment by Kahbazi** [comment]
I See. Thanks.

**Review by DylanDmitri — CHANGES_REQUESTED** [review]

**Comment by DylanDmitri** [comment]
Also, the build is complaining about reference assemblies, which is fixed by going to `/eng/scripts/` and running `.\GenerateRefenceAssemblies.ps1`

**Comment by anurse** [comment]
https://github.com/aspnet/AspNetCore/blob/master/docs/ReferenceAssemblies.md has some docs about how to generate the reference assemblies :). You can actually do it with an MSBuild command.

**Comment by DylanDmitri** [comment]
The MSBuild command was broken for me for a while so I didn't recommend it. It works now though!

File: src/Middleware/RequestThrottling/src/RequestThrottlingMiddleware.cs
Comment (Kahbazi) [reply src/Middleware/RequestThrottling/src/RequestThrottlingMiddleware.cs]: Do you mean that I should only store `OnRejected`?

File: src/Middleware/RequestThrottling/test/MiddlewareTests.cs
Comment (Kahbazi) [reply src/Middleware/RequestThrottling/test/MiddlewareTests.cs]: There are other tests which don't have OrTimeout after await. Should I add it to those too?

File: src/Middleware/RequestThrottling/src/RequestThrottlingMiddleware.cs
Comment (halter73) [inline src/Middleware/RequestThrottling/src/RequestThrottlingMiddleware.cs]: What if this throws? As it turns out, `_requestQueue.Release()` never gets called meaning the server will eventually stop accepting any new requests. This is ultimately the fault of not putting the start of the try/finally block in this method immediately after the call to `_requestQueue.TryEnterQueueAsync()`, but that's not the only thing that should be fixed.

Any time we call user-specified callbacks, we always wrap the call in a generic try/catch that swallows and logs any user-thrown exceptions. You can see an example of this where [Kestrel invokes the middleware pipeline.](https://github.com/aspnet/AspNetCore/blob/81f2d46660ec6ea92549edd55486de78a7597c58/src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs#L628-L651)

Comment (Tratcher) [reply src/Middleware/RequestThrottling/src/RequestThrottlingMiddleware.cs]: Kestrel's exception handling is special, it's at the top of the stack. A user callback throwing on the main request thread is fine, it gets logged by the server or exception handler and becomes a 500. No need to do anything extra so long as it the middleware state is maintained.

Comment (Tratcher) [reply src/Middleware/RequestThrottling/src/RequestThrottlingMiddleware.cs]: In fact, swallowing the exception could lead to false 200 responses.

Comment (Kahbazi) [reply src/Middleware/RequestThrottling/src/RequestThrottlingMiddleware.cs]: So I don't need to change anything, right?

Comment (DylanDmitri) [reply src/Middleware/RequestThrottling/src/RequestThrottlingMiddleware.cs]: With the first part of your comment @halter73, the code sure is super misleading. Also when implementing the other queuing strategy (that rejects at dequeue time) I needed to refactor this method in middleware to deal with the new flow, and then refactored it back after putting that on hold. Basically it's pretty fragile and I need to go back and work on it.

Comment (halter73) [reply src/Middleware/RequestThrottling/src/RequestThrottlingMiddleware.cs]: > In fact, swallowing the exception could lead to false 200 responses.

That would need to be accounted for of course. That's one reason why I recommended adding a throwing OnRejected test which we should add anyway, but it might not make too much sense to do so until we add more functional testing to verify we get a 503/500 instead of a 200 response.

After in person discussion with @Tratcher, we decided that since this exception will still get thrown from the main middleware pipeline, and that should still get logged with a helpful StackTrace, we're fine not catching here.

I personally still think it's more correct to respond with a 503 instead of a 500 and swallow/log the exception, but I can understand not wanting to add the complexity.

> So I don't need to change anything, right?

No. Hopefully the above explains the reasoning.

**Comment by DylanDmitri** [comment]
Alright @Kahbazi there's one last change and then it's all good to go! This is adding the unit test that Stephen mentioned above. 

Throw an exception during the `OnRejected` callback, and verify that the server replies appropriately.

File: src/Middleware/RequestThrottling/test/MiddlewareTests.cs
Comment (Kahbazi) [inline src/Middleware/RequestThrottling/test/MiddlewareTests.cs]: @halter73 @DylanDmitri I'm not sure if I quite understand the assertion part of the test that is needed. Is this what you had in mind?

Comment (DylanDmitri) [reply src/Middleware/RequestThrottling/test/MiddlewareTests.cs]: looks good!

File: src/Middleware/RequestThrottling/test/MiddlewareTests.cs
Comment (halter73) [inline src/Middleware/RequestThrottling/test/MiddlewareTests.cs]: This is a good start, but can we expand this test a little? I recommend the following.

1. Make `maxConcurrentRequests` equal to 1
2. Pass in a `next` RequestDelegate that returns the Task from a TCS (TaskCompletionSource) that you define at the top of the test.
3. Invoke the middleware and store the returned Task without awaiting it.
4. Invoke the middleware a second time and assert the DivideByZeroException like you're already doing.
5. Complete the TCS used by the `next`.
6. Assert that the Task returned from the first middleware invocation completes successfully.
7. Invoke the middleware a third time and Assert that it completes successfully since there's no other active requests and the Task returned by `next` is now completed.

You can also assert the `ActiveRequestCount` values like @DylanDmitri suggested in a previous comment.

Comment (DylanDmitri) [reply src/Middleware/RequestThrottling/test/MiddlewareTests.cs]: Another behavior change: we are afraid that people will overwrite `OnRejected` to log a message, and then the server will return an http 200.

The fix here is having the middleware set the statuscode to 503, and then invoking the user provided callback. The user can still set custom status codes, but by default it 503s.

Also, this means that the default `OnRejected` (in the options file) is now an empty lambda, since its current functionality gets moved to the middleware.

**Review by halter73 — APPROVED** [review]
Thanks @Kahbazi!

File: src/Middleware/RequestThrottling/src/RequestThrottlingOptions.cs
Comment (halter73) [inline src/Middleware/RequestThrottling/src/RequestThrottlingOptions.cs]: We still need add a doc comment for this though.

Comment (DylanDmitri) [reply src/Middleware/RequestThrottling/src/RequestThrottlingOptions.cs]: maybe "Optional callback; processes requests rejected by the middleware."

**Review by DylanDmitri — APPROVED** [review]

File: src/Middleware/RequestThrottling/src/RequestThrottlingOptions.cs
Comment (halter73) [reply src/Middleware/RequestThrottling/src/RequestThrottlingOptions.cs]: Everything has defaults, so I don't think we should bother pointing out it's optional. How about this?

"A <see cref="RequestDelegate"/> that handles requests rejected by this middleware. If it doesn't modify the response, an empty 503 response will be written."

Comment (halter73) [reply src/Middleware/RequestThrottling/src/RequestThrottlingOptions.cs]: Sorry to make you iterate so much @Kahbazi. @DylanDmitri I'm fine with you merging this as is and fixing up the doc comment later.

Comment (Kahbazi) [reply src/Middleware/RequestThrottling/src/RequestThrottlingOptions.cs]: no, it's ok. I can change it again when you decided what you want it to be. I think maybe "A callback that handles rejected request by this middleware." is good enough since 503 status code is now part  of the middleware and not OnRejected.

Comment (halter73) [reply src/Middleware/RequestThrottling/src/RequestThrottlingOptions.cs]: I do like making it clear that it's a RequestDelegate that can modify the response that is chooses to. I just realized my last comment wasn't formatted properly so you couldn't see the `<see cref="RequestDelegate"/>`'s. It was meant to be:

`"A <see cref="RequestDelegate"/> that handles requests rejected by this middleware. If it doesn't modify the response, an empty 503 response will be written."`

@DylanDmitri @Kahbazi Does this seem reasonable to you?

Comment (DylanDmitri) [reply src/Middleware/RequestThrottling/src/RequestThrottlingOptions.cs]: Stephen's description is better and more professional

**Pull request merged**
